pkg/transform/client.go

```diff
@@ -0,0 +1,306 @@
+// will be returned.
+//
+// If no names are provided, the full schema for the module is returned.
+// Otherwise, the resulting schema contains only the named elements and all of
+// their dependencies. This is enough information for the caller to construct
+// a dynamic message for any requested message types or to dynamically invoke
+// an RPC for any requested methods or services.
+func IncludeTypes(types ...string) ClientBuilder {
+	return func(c *Client) {
+		c.types = types
+	}
+}
+
+// FromFormat requires the format of the input data and an option to discard
+// unknown values. If true, any unresolvable fields in the input are discarded.
+// For formats other than FORMAT_BINARY, this means that the operation will
+// fail if the input contains unrecognized field names. For FORMAT_BINARY,
+// unrecognized fields can be retained and possibly included in the reformatted
+// output (depending on the requested output format).
+// TODO: supply logic in constructor or support the user through this flow
+func FromFormat(
+	format registryv1alpha1.Format,
+	discardUnknown bool,
+) ClientBuilder {
+	return func(c *Client) {
+		c.inputFormat = format
+		c.discardUnknown = discardUnknown
+	}
+}
+
+// Exclude configures the schema that is fetched from the schema service,
+// providing 2 configurable options:
+// `excludeCustomOptions` - If true, the returned schema will not include
+// extension definitions for custom options that appear on schema elements.
+// When filtering the schema based on the given element names, options on all
+// encountered elements are usually examined as well. But that is not the case
+// if excluding custom options.
+// `excludeKnownExtensions` - If true, the returned schema will not include
+// known extensions for extendable messages for schema elements. If
+// exclude_custom_options is true, such extensions may still be returned if the
+// applicable descriptor options type is part of the requested schema.
+//
+// These flags are ignored if `IncludeTypes()` is empty as the entire schema is
+// always returned in that case.
+func Exclude(excludeCustomOptions, excludeKnownExtensions bool) ClientBuilder {
+	return func(c *Client) {
+		c.excludeCustomOptions = excludeCustomOptions
+		c.excludeKnownExtensions = excludeKnownExtensions
+	}
+}
+
+// IfNotCommit is a commit that the client already has cached. So if the
+// given module version resolves to this same commit, the server should not
+// send back any descriptors since the client already has them.
+// This allows a client to efficiently poll for updates: after the initial RPC
+// to get a schema, the client can cache the descriptors and the resolved
+// commit. It then includes that commit in subsequent requests in this field,
+// and the server will only reply with a schema (and new commit) if/when the
+// resolved commit changes.
+func IfNotCommit(in string) ClientBuilder {
+	return func(c *Client) {
+		c.ifNotCommit = in
+	}
+}
```

Comment:
I don't think we'd expose this. Instead, this would be part of the client's internal cache management.

On first call, it retrieves a schema and version from the cache; if present, it includes that last seen version as this field in the RPC request.

On subsequent calls, it always includes the last seen version (and stores that version and new schema when it changes).